src/toil/job.py

```diff
@@ -3758,6 +3826,181 @@ def getUserScript(self):
+                return non_streaming_import.rv()
+            else:
+                raise
+
+
+class ImportsJob(Job):
+    """
+    Job to organize and delegate files to individual WorkerImportJobs.
+
+    For the CWL/WDL runners, this is only used when runImportsOnWorkers is enabled
+    """
+
+    def __init__(
+        self,
+        file_to_data: Dict[str, FileMetadata],
+        import_workers_threshold: ParseableIndivisibleResource,
+        **kwargs: Any,
+    ):
+        """
+        Job to take the inputs from the WDL workflow and import them on a worker instead of a leader. Assumes all local and cloud files are accessible.
+
+        This class is only used when runImportsOnWorkers is enabled.
+        """
+        super().__init__(local=True, **kwargs)
+        self._file_to_data = file_to_data
+        self._import_workers_threshold = import_workers_threshold
+
+    def run(
+        self, file_store: AbstractFileStore
+    ) -> Tuple[Promised[Dict[str, FileID]], Dict[str, FileMetadata]]:
+        """
+        Import the workflow inputs and then create and run the workflow.
+        :return: Tuple of a mapping from the candidate uri to the file id and a mapping of the source filenames to its metadata. The candidate uri is a field in the file metadata
+        """
+        import_workers_threshold = self._import_workers_threshold
+        file_to_data = self._file_to_data
+        # Run WDL imports on a worker instead
+
+        individual_files = []
+        batch_files = []
+        # Separate out files that will be imported individually and in a batch
+        for filename, (candidate_uri, parent_id, size) in file_to_data.items():
+            if size < import_workers_threshold:
+                batch_files.append(filename)
+            else:
+                individual_files.append(filename)
+
+        # Create individual import jobs
+        import_jobs = []
+        for file in individual_files:
+            filesize = file_to_data[file][2]
+            candidate_uri = file_to_data[file][0]
+            import_jobs.append(WorkerImportJob([candidate_uri], disk_size=filesize))
+
+        # create multiple batch jobs if necessary to not exceed the threshold
+        per_batch_files = []
+        per_batch_size = 0
+        # This list will hold lists of batched filenames
+        to_batch = []
+        while len(batch_files) > 0 or len(per_batch_files) > 0:
+            if len(batch_files) == 0 and len(per_batch_files) > 0:
+                to_batch.append(per_batch_files)
+                break
+
+            filename = batch_files[0]
+            # See if adding this to the queue will make the batch job too big
+            filesize = file_to_data[filename][2]
+            if per_batch_size + filesize >= import_workers_threshold:
+                # batch is too big now, store to schedule the batch
+                to_batch.append(per_batch_files)
+                # reset batching calculation
+                per_batch_size = 0
```

Comment:
Instead of relying on *not* popping the file and coming back around to it, you could start a new batch containing just it.